src/compartment.c

```diff
@@ -139,192 +142,302 @@ comp_from_elf(char* filename, struct ConfigEntryPoint* entry_points, size_t entr
+             comp_ehdr.e_shoff + comp_symtb_shdr.sh_link * sizeof(Elf64_Shdr));
+
+    // XXX Small trick here - we define the string table as a pointer
+    // to a char, even though it contains strings. This is done as
+    // string table indices are defined by char index, not string
+    // index. So when we index into them, we need to index at a char
+    // index.  However, since it actually contains indices, we can use
+    // string functions over dereferenced chars. This makes stuff hard
+    // to debug though...
+    char* comp_strtb = malloc(comp_strtb_hdr.sh_size);
+    do_pread((int) new_comp->fd, comp_strtb, comp_strtb_hdr.sh_size, comp_strtb_hdr.sh_offset);
+
+    Elf64_Sym* comp_symtb = malloc(comp_symtb_shdr.sh_size);
+    do_pread((int) new_comp->fd, comp_symtb, comp_symtb_shdr.sh_size, comp_symtb_shdr.sh_offset);
+
+    // Find symbols for entry_points
+    Elf64_Sym* ep_syms =
+        find_symbols((const char**) entry_points, entry_point_count,
+                     true, comp_symtb, comp_strtb, comp_symtb_shdr.sh_size);
+    for (size_t i = 0; i < entry_point_count; ++i)
+    {
+        struct entry_point* new_entry_point = malloc(sizeof(struct entry_point));
+        new_entry_point->fn_name = entry_points[i];
+        switch(new_comp->elf_type)
+        {
+            case ET_DYN:
+            {
+                new_entry_point->fn_addr = new_comp->base + ep_syms[i].st_value;
+                break;
+            }
+            case ET_EXEC:
             {
-                new_comp->relas[j] = comp_relas[j].r_offset;
-                // TODO double check
-                /*comp_relas[j].r_offset += new_comp->base;*/
-                /*uintptr_t old_plt_val = (uintptr_t) *((void**) comp_relas[j].r_offset);*/
-                /*old_plt_val += new_comp->base;*/
+                new_entry_point->fn_addr = ep_syms[i].st_value;
+                break;
             }
+            default:
+                assert(false);
+        }
+        new_comp->comp_fns[new_comp->entry_point_count] = new_entry_point;
+        new_comp->entry_point_count += 1;
+    }
+    free(ep_syms);
 
-            free(comp_relas);
+    // Find symbols for intercepts
+    char** intercept_names = calloc(intercept_count, sizeof(char*));
+    const char* so_plt_suffix = "@plt";
+    for (size_t i = 0; i < intercept_count; ++i)
+    {
+        if (new_comp->elf_type == ET_DYN)
+        {
+            size_t to_intercept_name_len = strlen(intercepts[i]) + strlen(so_plt_suffix) + 1;
+            intercept_names[i] = malloc(to_intercept_name_len);
+            strcpy(intercept_names[i], intercepts[i]);
+            strcat(intercept_names[i], so_plt_suffix);
+        }
+        else if (new_comp->elf_type == ET_EXEC)
+        {
+            intercept_names[i] = malloc(strlen(intercepts[i]) + 1);
+            strcpy(intercept_names[i], intercepts[i]);
         }
         else
+        {
+            assert(false && "Shouldn't get here");
```

Comment:
I tend to write this as:

```c
if (new_comp->elf_type == ET_DYN) { ... }
else {
  assert(new_comp->elf_type == ET_EXEC);
  ...
}
```

to cut the code down a bit.

Reply:
I'd like to keep it as before, in case we want to add more choices. Also, we might want to remove support entirely for static binaries, which would remove the `ET_EXEC` branch.